src/vs/workbench/api/common/extHostNotebook.ts

```diff
@@ -363,6 +376,88 @@ export class ExtHostNotebookController implements ExtHostNotebookShape {
 		return fileStats;
 	}
 
+	private async _validateWriteFile(uri: URI, options: files.IWriteFileOptions) {
+		// File system provider registered in Extension Host doesn't have unlock or atomic support
+		// Validate via file stat meta data
+		const stat = await this._extHostFileSystem.value.stat(uri);
+
+		// File cannot be directory
+		if ((stat.type & files.FileType.Directory) !== 0) {
+			throw new files.FileOperationError(localize('fileIsDirectoryWriteError', "Unable to write file '{0}' that is actually a directory", this._resourceForError(uri)), files.FileOperationResult.FILE_IS_DIRECTORY, options);
+		}
+
+		// File cannot be readonly
+		if ((stat.permissions ?? 0) & files.FilePermission.Readonly) {
+			throw new files.FileOperationError(localize('err.readonly', "Unable to modify read-only file '{0}'", this._resourceForError(uri)), files.FileOperationResult.FILE_PERMISSION_DENIED);
+		}
+
+		// Dirty write prevention
```

Comment:
@rebornix My feeling is that you actually only need this `Dirty write prevention`, but none of the checks above.